### PR TITLE
Feature/refactoring writers

### PR DESCRIPTION
**Pull request by UdayChokkam**
1. Moved common code from S3DirectoryWriter , DirectoryWriter, PlainFileWriter to a new class Writer
2. All tests   passed in local post refactoring

### CODE REVIEW

File: src/main/kotlin/app/batch/PlainFileWriter.kt [hunk 1]
```diff
@@ -1,28 +1,46 @@
...
+    private lateinit var outputFile: String
 
-    override fun write(items: MutableList<out String>) {
+    override fun writeData(encryptionResult: EncryptionResult, dataKeyResult: DataKeyResult) {
         logger.info("Appending to '$outputFile'.")
         val fw = FileWriter(outputFile, true)
-        items.forEach {
-            fw.write("$it\n")
-        }
+        // items.forEach {
```
Comment (mark-r-m) [inline src/main/kotlin/app/batch/PlainFileWriter.kt]: Please remove the comments

File: src/main/kotlin/app/batch/S3DirectoryWriter.kt [hunk 2]
```diff
@@ -1,102 +1,63 @@
...
-            if (batchSizeBytes + item.length > maxBatchOutputSizeBytes) {
-                writeOutput()
-            }
-            currentBatch.append(item)
-            batchSizeBytes += item.length
-        }
-    }
+                        private val cipherService: CipherService) : Writer<String>(keyService,cipherService){
+
+    @Value("\${aws.region}")
```
Comment (mark-r-m) [inline src/main/kotlin/app/batch/S3DirectoryWriter.kt]: The vars are at the bottom in other files, could you move them down?

File: src/main/kotlin/app/batch/S3DirectoryWriter.kt [hunk 3]
```diff
@@ -1,102 +1,63 @@
...
         // See also https://github.com/aws/aws-sdk-java
+        val fullFilePath = outputPath(++currentOutputFileNumber)
+        val fileBytes = encryptionResult.encrypted.toByteArray(StandardCharsets.US_ASCII)
         val bytesSize = fileBytes.size.toLong()
         logger.info("Writing file 's3://$s3BucketName/$fullFilePath' of '$bytesSize' bytes.")
 
         val inputStream = ByteArrayInputStream(fileBytes)
         val bufferedInputStream = BufferedInputStream(inputStream)
 
         // eu-west-1 -> EU_WEST_2 (i.e tf style to enum name)
```
Comment (mark-r-m) [inline src/main/kotlin/app/batch/S3DirectoryWriter.kt]: Could this comment be moved to be next to the region var?

File: src/main/kotlin/app/batch/Writer.kt [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+import java.io.OutputStream
+import java.nio.charset.StandardCharsets
+import java.nio.file.Files
+import java.nio.file.Path
+
+
+abstract class Writer<String>(private val keyService: KeyService,
+                              private val cipherService: CipherService) : ItemWriter<String> {
+
+    @Value("\${output.batch.size.max.bytes}")
```
Comment (mark-r-m) [inline src/main/kotlin/app/batch/Writer.kt]: The vars are at the bottom in other files, could you move them down?

File: src/main/kotlin/app/batch/S3DirectoryWriter.kt [hunk 5]
```diff
@@ -1,102 +1,63 @@
...
+    @Value("\${aws.region}")
+    private var region: kotlin.String = "eu-west-1"
+
+    @Value("\${s3.bucket}")
+    private lateinit var s3BucketName: kotlin.String // i.e. "1234567890"
+
+    @Value("\${s3.prefix.folder}")
+    private lateinit var s3PrefixFolder: kotlin.String //i.e. "mongo-export-2019-06-23"
+
+    val updated_region = region.toUpperCase().replace("-", "_")
```
Comment (mark-r-m) [inline src/main/kotlin/app/batch/S3DirectoryWriter.kt]: Camel case please

File: src/main/kotlin/app/batch/S3DirectoryWriter.kt [hunk 6]
```diff
@@ -1,102 +1,63 @@
...
-            batchSizeBytes += item.length
-        }
-    }
+                        private val cipherService: CipherService) : Writer<String>(keyService,cipherService){
+
+    @Value("\${aws.region}")
+    private var region: kotlin.String = "eu-west-1"
+
+    @Value("\${s3.bucket}")
+    private lateinit var s3BucketName: kotlin.String // i.e. "1234567890"
```
Comment (mark-r-m) [inline src/main/kotlin/app/batch/S3DirectoryWriter.kt]: Fyi you don't need the `kotlin.` in kotlin code :-)

File: src/main/kotlin/app/batch/PlainFileWriter.kt (hunk 1 above)
Comment (UdayChokkam) [reply src/main/kotlin/app/batch/PlainFileWriter.kt]: Thanks Mark , i have incorporated all  review comments.

File: src/main/kotlin/app/batch/DirectoryWriter.kt [hunk 7]
```diff
@@ -17,101 +16,49 @@ import java.nio.file.Paths
...
-            this.currentBatch = StringBuilder()
-            this.batchSize = 0
+
+    override fun writeData(encryptionResult: EncryptionResult,dataKeyResult: DataKeyResult) {
+        val dataPath = outputPath(++currentOutputFileNumber)
+        logger.info("Processing file number '%06d' with batchSize='$batchSizeBytes'.".format(currentOutputFileNumber))
+        Files.write(dataPath,
+                encryptionResult.encrypted.toByteArray(StandardCharsets.US_ASCII))
+        logger.info("Wrote dataPath: '$dataPath'.")
+
```
Comment (mark-r-m) [inline src/main/kotlin/app/batch/DirectoryWriter.kt]: Delete a blank line?

File: src/main/kotlin/app/batch/PlainFileWriter.kt [hunk 8]
```diff
@@ -1,28 +1,39 @@
...
 
 @Component
 @Profile("outputToFile")
-class FileSystemWriter : ItemWriter<String> {
+class FileSystemWriter(private val keyService: KeyService,
+                       private val cipherService: CipherService) : Writer<String>(keyService, cipherService) {
+    override fun outputPath(number: Int): Path {
+        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+    }
+
```
Comment (mark-r-m) [inline src/main/kotlin/app/batch/PlainFileWriter.kt]: Delete a blank line, move privates down

File: src/main/kotlin/app/batch/Writer.kt [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    private fun chunkData(items: MutableList<out String>) {
+        items.map { "$it\n" }.forEach { item ->
+            if (batchSizeBytes + item.length > maxBatchOutputSizeBytes) {
+                writeOutput()
+            }
+            currentBatch.append(item)
+            batchSizeBytes += item.length
+        }
+    }
+
```
Comment (mark-r-m) [inline src/main/kotlin/app/batch/Writer.kt]: Delete a blank line

File: src/main/kotlin/app/batch/Writer.kt [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    protected var batchSizeBytes = 0
+
+    protected var currentOutputFileNumber = 0
+
+    @Value("\${directory.output}")
+    protected lateinit var outputDirectory: kotlin.String
+
+    @Value("\${source.table.name}")
+    protected lateinit var tableName: kotlin.String // i.e. "db.user.data"
+
```
Comment (mark-r-m) [inline src/main/kotlin/app/batch/Writer.kt]: Delete a blank line

File: src/main/kotlin/app/batch/Writer.kt [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+    abstract fun  writeData(encryptionResult: EncryptionResult, dataKeyResult: DataKeyResult)
```
Comment (mark-r-m) [inline src/main/kotlin/app/batch/Writer.kt]: align these? With a blank line between each. Perhaps put them aa the top as they are critical?

**Pull request merged**